ga4gh/converters.py

```diff
@@ -6,35 +6,236 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import collections
+
+import pysam
+
+import ga4gh.protocol as protocol
+
 
 class AbstractConverter(object):
     """
     Abstract base class for converter classes
     """
-    def __init__(self, request, outputStream):
-        self._request = request
-        self._outputStream = outputStream
+    def __init__(self, httpClient):
+        self._httpClient = httpClient
+
+
+##############################################################################
+# SAM
+##############################################################################
+
+
+class SamException(Exception):
+    """
+    Something that went wrong during converting a SAM file
+    """
 
 
 class SamConverter(AbstractConverter):
     """
     Converts a request to a SAM file
     """
-    def __init__(self, searchReadsRequest, outputStream):
-        super(SamConverter, self).__init__(
-            searchReadsRequest, outputStream)
+    def __init__(self, httpClient, searchReadsRequest, outputFile,
+                 binaryOutput):
+        super(SamConverter, self).__init__(httpClient)
+        self._searchReadsRequest = searchReadsRequest
+        self._outputFile = outputFile
+        self._binaryOutput = binaryOutput
 
     def convert(self):
-        raise NotImplementedError()
+        header = self._getHeader()
+        targetIds = self._getTargetIds(header)
+        # pysam can't write to file streams (except for stdout)
+        # http://pysam.readthedocs.org/en/latest/usage.html#using-streams
+        if self._binaryOutput:
+            flags = "wb"
+        else:
+            flags = "wh"  # h for header
+        if self._outputFile:
+            fileString = self._outputFile
+        else:
+            fileString = "-"
+        alignmentFile = pysam.AlignmentFile(
+            fileString, flags, header=header.toDict())
+        iterator = self._httpClient.searchReads(self._searchReadsRequest)
+        for read in iterator:
+            alignedSegment = SamLine.toAlignedSegment(read, targetIds)
+            alignmentFile.write(alignedSegment)
+        alignmentFile.close()
+
+    def _getHeader(self):
+        header = SamHeader()
```

Comment:
I don't see much value in a `SamHeader` class here right now. Just make the dict directly here and return it. (If we need something more sophisticated in future then we should of course have a class)